### PR TITLE
Updates to Ruby Section

### PR DESCRIPTION
**Pull request by mattsinc**
Two (non-merge) commits here that update the Ruby section:

1.  First commit is various small fixes, clarifications, and additions throughout the Ruby section.
2.  Second commit is updates to the debugging section (SLICC) to add more information about protocol traces.

Matt

### CODE REVIEW

File: part3/MSI/debugging.rst [hunk 1]
```diff
@@ -61,6 +62,59 @@ Once you have your protocol working with the random tester, you can move on to u
 It is likely that real applications will expose even more bugs in the protocol.
 If at all possible, it is much easier to debug your protocol with the random tester than with real applications!
 
+Understanding Protocol Traces
+=============================
+
+Unfortunately, despite extensive effort to catch bugs in them, coherence protocols (even heavily tested ones) will have sometimes have bugs.
```
Comment (powerjg) [inline part3/MSI/debugging.rst]: have sometimes have -> will have (there are always bugs ;))

File: part3/MSI/debugging.rst [hunk 2]
```diff
@@ -61,6 +62,59 @@ Once you have your protocol working with the random tester, you can move on to u
...
 
+Understanding Protocol Traces
+=============================
+
+Unfortunately, despite extensive effort to catch bugs in them, coherence protocols (even heavily tested ones) will have sometimes have bugs.
+Sometimes these bugs are relatively simple fixes, while other times the bugs will be very insidious and difficult to track down.
+In the worst case, the bugs will manifest themselves as deadlocks: bugs that literally prevent the application from making progress.
+Another, similar problem is livelocks: where the program runs forever because there is a cycle somewhere in the system.
+Whenever livelocks or deadlocks occur, the next thing to do is generate a protocol trace.
+Traces print a running list of every transition that is happening in the memory system -- memory requests starting and completing, L1 and directory transitions, etc.
```
Comment (powerjg) [inline part3/MSI/debugging.rst]: Can you insert a "real" em-dash here (—)? Apparently that' what restructuredtext expects.

File: part3/MSI/debugging.rst [hunk 3]
```diff
@@ -61,6 +62,59 @@ Once you have your protocol working with the random tester, you can move on to u
...
+Thus, the above snippet is showing what was happening in the memory system between ticks 4541 and 5327.
+In this snippet, all of the requests are coming from L1Cache-0 (core 0) and going to Directory-0 (the first bank of the directory).
+During this time, we see several memory requests and state transitions for the cache line 0x4ac0, both at the L1 caches and the directory.
+For example, in tick 5322, the core executes a store to 0x4ac0.
+However, it currently does not have that line in Modified in its cache (it is in Shared after the core loaded it from ticks 4641-4674), so it needs to request ownership for that line from the directory (which receives this request in tick 5327).
+While waiting for ownership, it transitions from S (Shared) to SM_AD (a transient state -- was in S, going to M, waiting for Ack and Data).
+
+To add a print to the protocol trace, you will need to add a print with these fields with the ProtocolTrace flag.
+For example, if you look at ``src/mem/ruby/system/Sequencer.cc``, you can see where the ``Seq               Begin`` and ``Seq                Done`` trace prints come from (search for ProtocolTrace).
+
```
Comment (powerjg) [inline part3/MSI/debugging.rst]: This is great! Thanks for the added detail, Matt!

**Comment by mattsinc** [comment]
No problem, I spell check as I'm making changes because I don't trust myself to not make spelling mistake, so the changes are really being done as I'm making other changes.

I am closing this request, because I made a different one that reflects your fixes and requests above.

**Pull request closed without merging**